From c3be32854048c917e04b736692dcaa0130e464cb Mon Sep 17 00:00:00 2001
From: Don Zickus <dzickus@redhat.com>
Date: Tue, 16 Sep 2014 17:02:49 -0400
Subject: [usb] wusbcore: allow wa_xfer_destroy to clean up partially constructed xfers

Message-id: <1410887463-20674-98-git-send-email-dzickus@redhat.com>
Patchwork-id: 95276
O-Subject: [RHEL7 PATCH 097/591] usb: wusbcore: allow wa_xfer_destroy to clean up partially constructed xfers
Bugzilla: 1110939
RH-Acked-by: Tony Camuso <tcamuso@redhat.com>
RH-Acked-by: Prarit Bhargava <prarit@redhat.com>

https://bugzilla.redhat.com/show_bug.cgi?id=1110939

commit d993670ca97f646db1ef9b345e78ecfd3d6f0143
Author: Thomas Pugliese <thomas.pugliese@gmail.com>
Date:   Thu Sep 26 14:08:13 2013 -0500

    usb: wusbcore: allow wa_xfer_destroy to clean up partially constructed xfers

    If __wa_xfer_setup fails, it can leave a partially constructed wa_xfer
    object.  The error handling code eventually calls wa_xfer_destroy which
    does not check for NULL before dereferencing xfer->seg which could cause
    a kernel panic.  This change also makes sure to free xfer->seg which was
    being leaked for all transfers before this change.

    Signed-off-by: Thomas Pugliese <thomas.pugliese@gmail.com>
    Signed-off-by: Greg Kroah-Hartman <gregkh@linuxfoundation.org>
---

Signed-off-by: Jarod Wilson <jarod@redhat.com>

diff --git a/drivers/usb/wusbcore/wa-xfer.c b/drivers/usb/wusbcore/wa-xfer.c
index 47cbfdd..d2c7b2b 100644
--- a/drivers/usb/wusbcore/wa-xfer.c
+++ b/drivers/usb/wusbcore/wa-xfer.c
@@ -178,9 +178,15 @@ static void wa_xfer_destroy(struct kref *_xfer)
  if (xfer->seg) {
   unsigned cnt;
   for (cnt = 0; cnt < xfer->segs; cnt++) {
-   usb_free_urb(xfer->seg[cnt]->dto_urb);
-   usb_free_urb(&xfer->seg[cnt]->tr_urb);
+   if (xfer->seg[cnt]) {
+    if (xfer->seg[cnt]->dto_urb) {
+     kfree(xfer->seg[cnt]->dto_urb->sg);
+     usb_free_urb(xfer->seg[cnt]->dto_urb);
+    }
+    usb_free_urb(&xfer->seg[cnt]->tr_urb);
+   }
   }
+  kfree(xfer->seg);
  }
  kfree(xfer);
 }
-- 
1.7.1